pkg/billing/model.go

```diff
@@ -23,3 +26,44 @@ type IncrementalEvent struct {
 	StopTime       time.Time `json:"stop_time"`
 	Value          int       `json:"value"`
 }
+
+type enrichable interface {
+	setEventType()
+	idempotencyKey() *string
+}
+
+// Enrich sets the event's Type and IdempotencyKey fields, so that users of this API don't need to
+// manually set them
+//
+// Enrich is already called by (*Batch).Add, but this is used in the autoscaler-agent's billing
+// implementation, which manually calls Enrich in order to log the IdempotencyKey for each event.
+func Enrich[E enrichable](batch *Batch, event E) E {
+	event.setEventType()
+
+	key := event.idempotencyKey()
+	if *key == "" {
+		*key = fmt.Sprintf("Host<%s>:ID<%s>:T<%s>", batch.c.hostname, uuid.NewString(), time.Now().Format(time.RFC3339))
```

Comment:
in the spirit of "encapsulation" it seems like this really just wants to be a method on batch because you have to look at it's private fields, and the only reason it can't be a method is the generic thing, which (I might be missing something subtle, but) I think is avoidable.

Reply:
Made the change.